src/Middleware/InitializeTenancyByRequestData.php

```diff
@@ -0,0 +1,64 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Stancl\Tenancy\Middleware;
+
+use Closure;
+use Illuminate\Http\Request;
+use Stancl\Tenancy\Exceptions\TenantCouldNotBeIdentifiedException;
+
+class InitializeTenancyByRequestData
+{
+    /** @var callable */
+    protected $onFail;
+
+    public function __construct(callable $onFail = null)
+    {
+        $this->onFail = $onFail ?? function ($e) {
+            throw $e;
+        };
+    }
+
+    /**
+     * Handle an incoming request.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @param  \Closure  $next
+     * @return mixed
+     */
+    public function handle($request, Closure $next)
+    {
+        if ($request->method() !== 'OPTIONS') {
+            try {
+                $this->parseTenant($request);
+            } catch (TenantCouldNotBeIdentifiedException $e) {
+                ($this->onFail)($e);
+            }
+        }
+
+        return $next($request);
+    }
+
+    protected function parseTenant(Request $request)
+    {
+        if (tenancy()->initialized) {
+            return;
+        }
+
+        $tenant = null;
+        if ($request->hasHeader('X-Tenant')) {
```

Comment:
change to a config value as not all wants a `X-Tenant` in there app.

Reply:
Alright, also changed the possible value for the query parameter